katsdpimageutils/primarybeamcorrection/primary_beam_correction.py

```diff
@@ -1,188 +1,219 @@
-    # Calling the radial_offset function
-    separation_rad = radial_offset(phase_center, image_wcs, row, col)
-    # Calling the cosine_power_pattern function
+    phase_center, image_wcs = get_position(path)
+    separation_rad = radial_offset(phase_center, image_wcs, pixcrd)
     beam_list = cosine_power_pattern(separation_rad, c_freq)
-    return beam_list
+    return beam_list, data
 
 
-def mad(data):
-    """
-    Calculating median absolute deviation (MAD).
+def standard_deviation(data):
+    """Compute the median and the estimate of the standard deviation
+    based on the median absolute deviation (MAD).
     """
     MAD_TO_SD = 1.4826
     med = np.nanmedian(data)
     dev = np.abs(data - med)
     return med, MAD_TO_SD * np.nanmedian(dev)
 
 
-def standard_deviation(data):
+def inverse_variance(data):
+    """ Calculate the inverse variance. Reject pixels more than 5 sigma from
+    the median and reject all zeros until either no more pixels are rejected or
+    a maximum  of 50 iterations is reached.
     """
-    Calculating the standard deviation of each frequency plane using MAD. Rejecting pixels more
-    than 5 sigma from the mean until either no more pixels are rejected or a maximum  of 50
-    iterations is reached.
-    Returns: Weights per each frequency plane
-    """
-    diff = 1
-    i = 0
     data = data[data != 0.0]
     if len(data) == 0:
         return float(0.0), float(0.0)
-    med, sd = mad(data)
-    while diff > 0:
-        i += 1
-        if i > 50:
-            return 1/(sd)**2
+    med, sd = standard_deviation(data)
+    for i in range(50):
         old_sd = sd
-        old_len = len(data)
         cut = np.abs(data - med) < 5.0 * sd
         if np.all(~cut):
             return 1/(sd)**2
         data = data[cut]
-        med, sd = mad(data)
-        diff = old_len - len(data)
+        med, sd = standard_deviation(data)
         if sd == 0.0:
             return 1/(old_sd)**2
+        if i == 49:
+            return 1/(sd)**2
     return 1/(sd)**2
 
 
-def weighted_average(arr, weights, mask_image):
+def weighted_average(arr, weights):
+    """Compute weighted average of all the frequency planes.
     """
-    Computing weighted average of all the frequency planes.
-    """
-    masked_arr = np.ma.MaskedArray(arr, mask=mask_image)
-    wt_average = np.average(masked_arr, weights=weights, axis=0)
-    return wt_average.data
+    wt_average = np.average(arr, weights=weights, axis=0)
+    return wt_average
 
 
-def primary_beam_correction(beam_pattern, path):
-    """
-    Correcting for the primary beam effects.
+def primary_beam_correction(beam_pattern, raw_image, px_cut=0.1):
+    """Correct the effects of primary beam.
+    
+    Parameters
+    ----------
+    beam_pattern: numpy array
+        Beam battern made from the Cosine-squared power approximation
+        from Mauch et al. (2020).
+    path: astropy object
+        First element of the HDU list
```

Comment:
Presumably this function would work for any ImageHDU, wherever it is in the list.

Reply:
True that I made it more generally now.